### DISCUSSION TITLE
Running Japa tests on subdomain (tenant)

### DISCUSSION THREAD

**Discussion by carlmathisen** [adonisjs/core#3712] [Help] [answered]
Hi everyone,

I have a route group that I want to run Japa tests towards.

The issue I have is that all of my routes are behind a domain configuration, such as `tenant.mycompany.com`, or locally, `tenant.localhost:3333`.

When I run `node ace test`, it will spin up a dedicated server to run tests towards:

```shell
$ node ace test
[ info ]  running tests...
{ ... "msg":"started server on 0.0.0.0:62054"}
```

When my HTTP tests are being executed using `client.post('/my/endpoint')`, the tests will automatically run against `http://0.0.0.0:62054/my/endpoint`. It will run without my tenant subdomain, and fail since the routes won't be triggered.

The test requests will return `message: 'E_ROUTE_NOT_FOUND: Cannot POST:/storage/test-project'`

I tried to configure `apiClient` in `tests/bootstrap.ts` so that `apiClient` is configured to run against a tenant URL:

`tests/bootstrap.ts`
```ts
...
import { domain } from 'Config/...' // domain is set to 'localhost'
const testUrl = `http://test.${domain}`

export const plugins: Config['plugins'] = [assert(), runFailedTests(), apiClient(testUrl)]
```

If i then run `node ace test`, the tests don't run towards the correct address:

```shell
$ node ace test
[ info ]  running tests...
{ ... "msg":"started server on 0.0.0.0:62792"}

...

connect ECONNREFUSED 127.0.0.1:80

   at TCPConnectWrap.afterConnect [as oncomplete] net.js:1159
```

I guess this makes sense, because the test server is running on a random port, while the tests are pointing towards my `apiClient` configuration. Next up would be to figure out which port the test server is running on.

But before I venture into that rabbit hole I wanted to ask here -- Am I going down the correct path? 😅 

Thanks!
Carl

---

See project code snippets below:

`start/routes.ts`
```ts
import Route from '@ioc:Adonis/Core/Route'
import Config from '@ioc:Adonis/Core/Config'

// this is just a config variable that allows me to set a dynamic domain, such as "mycompany.com"
const domain = Config.get('els.domain')

Route.group(() => {
  Route.post(':projectId', 'StorageController.store')
})
  // example: tenant.mycompany.com
  .domain(`:tenant.${domain}`)
  .prefix('storage')
```

`tests/functional/storage.spec.ts`
```ts
import { test } from '@japa/runner'

test.group('Storage', () => {
  test('Store payload', async ({ client }) => {
    const response = await client
      .post('storage/test-project')
      .json({
        metadata: {
          message: 'hi mom'
        }
      })
    response.dump()
  })
})

```

**Comment by carlmathisen** [comment, marked as answer]
Ok, I got it working by adding the port to my `apiClient` configuration:

`tests/bootstrap.ts`
```ts
...
const port = Env.get('PORT')
const testUrl = `http://test.${domain}:${port}`
export const plugins: Config['plugins'] = [assert(), runFailedTests(), apiClient(testUrl)]
...
```

I hope this is useful for someone else too. Let me know if there's a better way. Thanks!

**Reply by victorLessa** [reply]
Is it possible to change the baseUrl dynamically? I have tests for routes with tenant and routes without tenant.

**Reply by NullVoxPopuli** [reply]
I have a need for changing the subdomain per test as well

**Reply by NullVoxPopuli** [reply]
I figured it out -- I just removed the url from apiClient, and I now specify the full path in each test

**Reply by RomainLanz** [reply]
Since it wasn't clear from the documentation, may you send a PR to improve it?
